### PR TITLE
Restart test executions functionality

### PR DESCRIPTION
**Pull request by omar-selo**
TO PR for resolving [RTW-303](https://warthogs.atlassian.net/browse/RTW-303). (there should still be a small PR on hwcert-jenkins-jobs to complement this)

Changes:
- Adds a new table named `test_execution_rerun_request`
- Adds a new endpoint `/v1/test-executions/reruns` that you can post to and get
- API returns an additional field with test executions named `is_rerun_requested`
- Delete and recreation of `poetry.lock` file to solve [local dev environment not working](https://chat.canonical.com/canonical/pl/wec55ggwqfrmtffcsd19cynqaw). This required a new db migration and small modifications to a function in `repository.py`
- Add a new script `test_executions_rerunner.py` that will be used in jenkins to actually perform those reruns (this way jenkins credentials stay in jenkins)
- When `PUT /v1/test-executions/start-test` is called delete existing rerun request
- Added rerun button to the frontend. This button is disabled if a rerun was requested already

[Screencast from 2024-05-01 16-53-58.webm](https://github.com/canonical/test_observer/assets/4087200/f0f02ec6-7efa-4b73-bdd6-a309b872e90a)

[RTW-303]: https://warthogs.atlassian.net/browse/RTW-303?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backend/migrations/versions/2024_04_30_1029-33c0383ea9ca_rename_types_due_to_sqlalchemy_alembic_.py
Comment (omar-selo) [inline backend/migrations/versions/2024_04_30_1029-33c0383ea9ca_rename_types_due_to_sqlalchemy_alembic_.py:1]: This migration was required after recreating `poetry.lock`

File: backend/test_observer/data_access/repository.py
Comment (omar-selo) [inline backend/test_observer/data_access/repository.py:1]: These changes were required after recreating `poetry.lock`

File: backend/test_observer/controllers/test_executions/reruns.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    Stage,
+    TestExecution,
+    TestExecutionRerunRequest,
+)
+from test_observer.data_access.repository import get_or_create
+from test_observer.data_access.setup import get_db
+
+from .models import PendingRerun, RerunRequest
+
+router = APIRouter()
```
Comment (andrejvelichkovski) [inline backend/test_observer/controllers/test_executions/reruns.py:17]: Can we add `tags=["re-runs"]` to separate these in a different category in the docs view?

File: frontend/lib/providers/artefact_builds.dart [hunk 2]
```diff
@@ -70,4 +70,25 @@ class ArtefactBuilds extends _$ArtefactBuilds {
...
+            te.copyWith(isRerunRequested: true)
+          else
+            te,
+      ];
+
+      newArtefactBuilds.add(ab.copyWith(testExecutions: newTestExecutions));
+    }
+
+    state = AsyncData(newArtefactBuilds);
+  }
```
Comment (andrejvelichkovski) [inline frontend/lib/providers/artefact_builds.dart:93]: I remember implementing this `_getUpdatedArtefactBuild` ([link](https://github.com/canonical/test_observer/blob/baefb5009ca6dcffa60eb27cca76b01949b81413/frontend/lib/providers/artefact_builds.dart#L18)) which seems to have the same behaviour as the code here. Perhaps we can either re-use it or refactor that one to match this one?

**Review by andrejvelichkovski — DISMISSED** [review]
Excited to see this landed as it will be a really useful addition. I left two comments on how we can improve the code, but we can land this as it is.

File: backend/test_observer/controllers/test_executions/reruns.py (hunk 1 above)
Comment (omar-selo) [reply backend/test_observer/controllers/test_executions/reruns.py:17]: Ah keep forgetting about tags

Comment (omar-selo) [reply backend/test_observer/controllers/test_executions/reruns.py:17]: Hmm looks like if we add rerun tag then the docs will show the endpoint twice. Once for rerun tag and once for test-executions tag:
image

Not sure that's desirable. We could move this out of test-executions tag, but I prefer if it's there. So how about we do without rerun tag?

Comment (andrejvelichkovski) [reply backend/test_observer/controllers/test_executions/reruns.py:17]: We can do without the rerun tag. I didn't realise it is under the test-executions tag, and was thinking it is actually under the default tag.

**Review by andrejvelichkovski — APPROVED** [review]
LGTM, thanks for refactoring the common code in one!

**Pull request merged**